### DISCUSSION TITLE
Using WebDriverIO with third-party browsers

### DISCUSSION THREAD

**Discussion by lauromoura** [webdriverio/webdriverio#11880] [Q&A]
Hi, does WebdriverIO support using third-party browsers other than Chrome/Edge/Safari/Firefox? The browsers in question are based on WebKit, like Safari (e.g. GNOME's Epiphany and WPE's Cog).

I tried playing with the capabilities in the manual mode but looks like `browserName` must be one of these 4 (or closely related), with WebDriverIO refusing to start `remote()` if it's different. If I try using `safari` but changing other caps like binaries paths, the wrong `browserName` causes the capability negotiation to fail with the third-party browser service.

Before testing WebDriverIO, I've been using Selenium's python bindings, through its `RemoteWebDriver` classes, with the extra capabilities specific to the tested browsers like command line arguments. Does WebDriverIO have something similar to the `RemoteWebDriver`?

**Comment by christian-bromann** [comment]
It seems like these browser have their own drivers that you need to download and run beforehand. For example Epiphany uses a [WebKitGTK](https://github.com/SeleniumHQ/selenium/tree/master/py/selenium/webdriver/webkitgtk) driver that you can start on a certain port and have your WebdriverIO session point towards this endpoint, e.g.

```ts
// wdio.conf.js
export const config = {
  // ...
  hostname: 'localhost',
  port: 1234,
  // ...
}
```

I am not sure if Safaridriver allows to automate a WebKit instance. I can move this into an issue to have extend the docs around this topic.

**Reply by christian-bromann** [reply]
https://github.com/webdriverio/webdriverio/issues/11899 , any contributions or help would be appreciated.

**Comment by lauromoura** [comment]
Sorry for the delay responding. Setting the port and hostname properties made it work fine. Thanks for the reply! 

> It seems like these browser have their own drivers that you need to download and run beforehand. For example Epiphany uses a [WebKitGTK](https://github.com/SeleniumHQ/selenium/tree/master/py/selenium/webdriver/webkitgtk) driver that you can start on a certain port and have your WebdriverIO session point towards this endpoint, e.g.

Yes, WebKitGTK and WPEWebKit are ports of WebKit, and we have `WebKitWebDriver` and `WPEWebDriver`, respectively, inside WebKit's own repository.

> 
> I am not sure if Safaridriver allows to automate a WebKit instance.

Do you mean using Safaridriver to automate a generic WebKitBrowser? If so, I'm afraid it's not possible. From https://developer.apple.com/documentation/webkit/testing_with_webdriver_in_safari:  "Each safaridriver is capable of launching only the Safari version it’s associated with...".

> I can move this into an issue to have extend the docs around this topic.

I'll try to help with the documentation issue.
